### DISCUSSION TITLE
Query data and fill a form with that data

### DISCUSSION THREAD

**Discussion by ebisbe** [DamianOsipiuk/vue-query#190] [Q&A] [answered]
I'm trying to fill an Edit form with that from the vue-query. Should I use a watch to see when the `{ data }` is not undefined and copy that information to a new property? I've been searching info about how should be done but I haven't find anything relevant.

**Comment by DamianOsipiuk** [comment, marked as answer]
That depends on what UX you want to achieve and what form would you use (library or native elements).
Basically you have to options:
- use `v-if` to check when data is available and display some fallback otherwise, like a spinner (similar to examples)
- use `computed` to get data or fallback if data is empty, like empty object. Then you can just use computed value in the template without `v-if`

**Reply by ebisbe** [reply]
I used the `v-if` strategy but somehow I can't modify the data. It's like inmutable. I can write on the textfields but as soon as I leave the field it resets to the default data. In the console I get `Set operation on key "" failed: target is readonly.`. From the `useQuery` I'm returning a Class. I try it to remove the class but it didn't change the message.

Any idea?

**Reply by DamianOsipiuk** [reply]
Yes, values returned by the query are immutable, so you would not be able to mutate the cache state and that is by design.
In this case you should clone deep the data inside `computed` compostable and pass returned value to your form.

**Reply by ebisbe** [reply]
I tried to use a `watch` but could not making it work. There are no events that fire `onSuccess` if I use 
```
export function useTodoQuery(todoId, { enabled }) {
  return useQuery(
    ["todos", todoId],
    () => axios.get(`/todos/${todoId.value}`),
    {
      enabled,
      select: (todo) => todo.data,
    }
  );
}
```
right?
I saw that there is one in the useQuery but thats not returned.

**Reply by DamianOsipiuk** [reply]
I was thinking about something like:

```ts
const { data } = useTodoQuery(...)
// This is only for storage purpose
const newData = ref()
const formData = computed(() => {
    // if you already have something changed in the form
    // This will also prevent data loss when additional fetch happens when user is filling the form as it would not be overridden
    if (newData.value) {
      return newData.value;
    }
    // if form is pure and query already has data, copy it as a formData
    if (data.value) {
      // cloneDeep here to prevent mutations on nested objects
      newData.value = cloneDeep(data.value)
      return newData.value;
    }
    // if there is no data yet
    return {}
})
```

[Playground](https://sfc.vuejs.org/#eyJBcHAudnVlIjoiPHNjcmlwdCBzZXR1cD5cbmltcG9ydCB7IHJlYWN0aXZlLCBjb21wdXRlZCwgcmVmIH0gZnJvbSAndnVlJ1xuXG4vLyB1c2VRdWVyeSBoZXJlXG5jb25zdCBkYXRhID0gcmVmKHthOiAxLCBiOjIsYzozLCBkOiB7ZTogNn19KVxuLy8gVGhpcyBpcyBvbmx5IGZvciBzdG9yYWdlIHB1cnBvc2VcbmNvbnN0IG5ld0RhdGEgPSByZWYoKVxuY29uc3QgZm9ybURhdGEgPSBjb21wdXRlZCgoKSA9PiB7XG4gICAgLy8gaWYgeW91IGFscmVhZHkgaGF2ZSBzb21ldGhpbmcgY2hhbmdlZCBpbiB0aGUgZm9ybVxuICAgIC8vIFRoaXMgd2lsbCBhbHNvIHByZXZlbnQgZGF0YSBsb3NzIHdoZW4gYWRkaXRpb25hbCBmZXRjaCBoYXBwZW5zIHdoZW4gdXNlciBpcyBmaWxsaW5nIHRoZSBmb3JtIGFzIGl0IHdvdWxkIG5vdCBiZSBvdmVycmlkZGVuXG4gICAgaWYgKG5ld0RhdGEudmFsdWUpIHtcbiAgICAgIHJldHVybiBuZXdEYXRhLnZhbHVlO1xuICAgIH1cbiAgICAvLyBpZiBmb3JtIGlzIHB1cmUgYW5kIHF1ZXJ5IGFscmVhZHkgaGFzIGRhdGEsIGNvcHkgaXQgYXMgYSBmb3JtRGF0YVxuICAgIGlmIChkYXRhLnZhbHVlKSB7XG4gICAgICAvLyBjbG9uZURlZXAgaGVyZSB0byBwcmV2ZW50IG11dGF0aW9ucyBvbiBuZXN0ZWQgb2JqZWN0c1xuICAgICAgbmV3RGF0YS52YWx1ZSA9IHsuLi5kYXRhLnZhbHVlLCBkOiB7Li4uZGF0YS52YWx1ZS5kfX1cbiAgICAgIHJldHVybiBuZXdEYXRhLnZhbHVlO1xuICAgIH1cbiAgICAvLyBpZiB0aGVyZSBpcyBubyBkYXRhIHlldFxuICAgIHJldHVybiB7fVxufSlcblxuc2V0VGltZW91dCgoKSA9PiB7XG4gIGZvcm1EYXRhLnZhbHVlLmQuZSA9ICdmb28nXG4gIGZvcm1EYXRhLnZhbHVlLmEgPSAnYmFyJ1xufSwgMjAwMClcbiAgXG48L3NjcmlwdD5cblxuPHRlbXBsYXRlPlxuICA8cD5RdWVyeSBkYXRhOiB7eyBkYXRhIH19PC9wPlxuICA8cD5Gb3JtIGRhdGE6IHt7IGZvcm1EYXRhIH19PC9wPlxuPC90ZW1wbGF0ZT4iLCJpbXBvcnQtbWFwLmpzb24iOiJ7XG4gIFwiaW1wb3J0c1wiOiB7XG4gICAgXCJ2dWVcIjogXCJodHRwczovL3NmYy52dWVqcy5vcmcvdnVlLnJ1bnRpbWUuZXNtLWJyb3dzZXIuanNcIixcbiAgICBcInZ1ZS9zZXJ2ZXItcmVuZGVyZXJcIjogXCJodHRwczovL3NmYy52dWVqcy5vcmcvc2VydmVyLXJlbmRlcmVyLmVzbS1icm93c2VyLmpzXCJcbiAgfVxufSJ9)

**Reply by ebisbe** [reply]
it throws a `// eslint-disable-next-line vue/no-side-effects-in-computed-properties` but works. So thanks

**Reply by DamianOsipiuk** [reply]
You could also try something like this with watch which should not produce an error: [Playground](https://sfc.vuejs.org/#eyJBcHAudnVlIjoiPHNjcmlwdCBzZXR1cD5cbmltcG9ydCB7IHJlYWN0aXZlLCBjb21wdXRlZCwgcmVmLCB3YXRjaCwgcmVhZG9ubHkgfSBmcm9tICd2dWUnXG5cbi8vIHVzZVF1ZXJ5IGhlcmVcbmNvbnN0IGRhdGEgPSByZWYoKVxuLy8gV29ya2Fyb3VuZCBmb3IgVnVlIHdhdGNoIG5vdCB3b3JraW5nIHdpdGggcmVhZG9ubHkgcmVmc1xuY29uc3QgbmV3RGF0YSA9IGNvbXB1dGVkKCgpID0+IGRhdGEudmFsdWUpXG4vLyBZb3VyIGRhdGEgZ29lcyBoZXJlXG5jb25zdCBmb3JtRGF0YSA9IHJlZigpO1xud2F0Y2goKCkgPT4gISFuZXdEYXRhLnZhbHVlLCAoKSA9PiB7XG4gIC8vIENsb25lIGRlZXAgaGVyZVxuICBmb3JtRGF0YS52YWx1ZSA9IHsuLi5kYXRhLnZhbHVlLCBkOiB7Li4uZGF0YS52YWx1ZS5kfX1cbn0pXG5cbnNldFRpbWVvdXQoKCkgPT4ge1xuICBkYXRhLnZhbHVlID0gcmVhZG9ubHkoe2E6IDEsIGI6MixjOjMsIGQ6IHtlOiA2fX0pXG59LCAxMDAwKVxuICBcbnNldFRpbWVvdXQoKCkgPT4ge1xuICBkYXRhLnZhbHVlID0gJ2NoYW5nZWQnXG59LCAzMDAwKVxuICBcbnNldFRpbWVvdXQoKCkgPT4ge1xuICBmb3JtRGF0YS52YWx1ZS5kLmUgPSAnZm9vJ1xuICBmb3JtRGF0YS52YWx1ZS5hID0gJ2Jhcidcbn0sIDIwMDApXG48L3NjcmlwdD5cblxuPHRlbXBsYXRlPlxuICA8cD5RdWVyeSBkYXRhOiB7eyBkYXRhIH19PC9wPlxuICA8cD5Gb3JtIGRhdGE6IHt7IGZvcm1EYXRhIH19PC9wPlxuPC90ZW1wbGF0ZT4iLCJpbXBvcnQtbWFwLmpzb24iOiJ7XG4gIFwiaW1wb3J0c1wiOiB7XG4gICAgXCJ2dWVcIjogXCJodHRwczovL3NmYy52dWVqcy5vcmcvdnVlLnJ1bnRpbWUuZXNtLWJyb3dzZXIuanNcIixcbiAgICBcInZ1ZS9zZXJ2ZXItcmVuZGVyZXJcIjogXCJodHRwczovL3NmYy52dWVqcy5vcmcvc2VydmVyLXJlbmRlcmVyLmVzbS1icm93c2VyLmpzXCJcbiAgfVxufSJ9)
